### PR TITLE
Fix #16: tests.

### LINKED ISSUE #16: Most tests in this module don't actually test a anything
For example, the second one:

```js
test('fail if request body is already parsed', function(t) {
  var mockReq = httpMocks.createRequest({
    method: 'POST',
    headers: { },
    _body: true,
    _rawBody: {},
    on: function(eventName, callback) { }
  })

  var mockRes = httpMocks.createResponse({
    status: function(httpCode) {
      t.equal(httpCode, 400)
      return {
        json: function(input) {
          t.deepEqual(input, { status: 'failure', reason: 'The raw request body is not available.' })
        }
      }
    }
  })

  var nextInvocationCount = 0
  var mockNext = function() { nextInvocationCount++ }

  verifier(mockReq, mockRes, mockNext)

  t.equal(nextInvocationCount, 0)
  t.end()
})
```

It never invokes that `status` function or compare the equal json. The fix here is to rename `_rawBody` to `rawBody`, that is probably a typo, and to just compare the output of `mockRes`:

```js
test('fail if request body is already parsed', function(t) {
  var mockReq = httpMocks.createRequest({
    method: 'POST',
    headers: {},
    _body: true,
    rawBody: {}
  })

  var responseHttpCode;

  var mockRes = httpMocks.createResponse();

  var nextInvocationCount = 0
  var mockNext = function() { nextInvocationCount++ }

  verifier(mockReq, mockRes, mockNext)

  t.equal(mockRes.statusCode, 400)
  t.deepEqual(JSON.parse(mockRes._getData()), {
    reason: "The raw request body is not available.",
    status: "failure"
  });
  
  t.equal(nextInvocationCount, 0)
  t.end()
})
```

I started fixing others, however maybe there's abuse of mocks here IMO and actually mounting the middleware on express and invoking it would make tests much shorter and clearer?

### PR DESCRIPTION
**Pull request by dblock**
This fixes #16, the tests actually test that middleware functionality. 

I am not sure how to write an actual verifier passing test, it requires stubbing the verifier inside alexa-verifier somehow.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -9,7 +9,7 @@ var verifier = require('alexa-verifier')
 module.exports = function alexaVerifierMiddleware(req, res, next) {
   if (req._body && (typeof req.rawBody !== 'string')) {
     var er = 'The raw request body is not available.'
-    res.status(400).json({ status: 'failure', reason: er })
+    return res.status(400).json({ status: 'failure', reason: er })
```
Comment (dblock) [inline index.js:12]: In real life this aborts the request, but it's too hard to have that in tests with mocks. I think this doesn't hurt.

**Comment by tejashah88** [comment]
Unfortunately, it would be nearly impossible to make a passing case unless you were Amazon themselves. Logically, if we did have a pass case, then there is a bug with either this repo or the `alexa-verifier` repo, since that would be a severe security flaw.

One approach that we can take when making these tests is, "What are the possible ways to break the verification modules?"

**Comment by dblock** [comment]
I think replacing the actual check with a mock implementation is the way to go, but I wasn't able to actually accomplish it.

**Comment by tejashah88** [comment]
What I meant to say is that, there is no way to make a proper passing case with a mock. You'd basically need Amazon's servers to send a test request in order for the passing test to work.

I even checked the `alexa-verifier` module and there's [no passing cases](https://github.com/mreinstein/alexa-verifier/blob/master/test/index.js) found. IMO, the only working way of testing if it works is with an actual skill.

**Comment by dblock** [comment]
LMK if I need to do anything else here for you to merge it.

**Comment by tejashah88** [comment]
So I found the original section upon how exactly should any webservice verify Alexa requests [here](https://developer.amazon.com/public/solutions/alexa/alexa-skills-kit/docs/developing-an-alexa-skill-as-a-web-service#verifying-that-the-request-was-sent-by-alexa).

I've created a simple skill and was almost able to make a passing case:

image

Aside from bypassing an error that states that the request is too old, it works just fine. The trick is that the timestamp needs to be up to date and the signature needs to be generated based on the request body.

image

Another note is that the `signaturecertchainurl` is set to `https://s3.amazonaws.com/echo.api/echo-api-cert-4.pem` to avoid the `certificate expiration check failed` error.

**Comment by dblock** [comment]
If you look at the code in this PR it generates the current timestamp with `timestamp: new Date().getTime()`, if that's what you're asking.

**Comment by tejashah88** [comment]
What I mean is that that above code for the passing case I made only fails when the request is made more than 150 seconds ago (i.e. the request is too old). If I comment [that portion](https://github.com/mreinstein/alexa-verifier/blob/master/index.js#L63) out, the passing case actually works.

When Amazon sends the request, the request body is used as a basis for generating the `signature`. The `signature` itself is base64-encoded and decoding it results in an encrypted signature, which can only be decrypted with the public key that's embedded in the certificate downloaded from `signaturecertchainurl`. The final result is a hash value. Next, we'd generate a SHA-1 hash from the request body and compare it to the first hash we derived from.

**TL;DR**, Either we have to generate the `signature` dynamically since the timestamp will be different, or we need Amazon to somehow send a test request with the correct signature.

Here's some semi-pseudocode that would be used to generate the `signature`

```javascript
var requestBody = {...}

var signatureCert = download("https://s3.amazonaws.com/echo.api/echo-api-cert-4.pem")
var publicKey = extractPublicKey(signatureCert)

var sha1Hash = generateHash(requestBody, 'SHA-1')
var encryptedSignature = encrypt(sha1Hash, publicKey)
var finalSignature = base64encode(encryptedSignature)
```

**Comment by dblock** [comment]
Since you don't have a private key for the signature you cannot generate a valid one on the fly. That would defeat the whole purpose of signatures here, as anyone would be able to generate a request. However you can use an outdated signed request if you could fake time, I think you can stub `.getTime` with `sinon` and return a well known value for the test.

That said, while this discussion is interesting can we do things one after another? This PR fixes tests that otherwise don't test anything, can it be merged?

**Comment by tejashah88** [comment]
Yes, I'll merge it. We can open another PR/issue about it for the passing case in the meantime.

**Comment by dblock** [comment]
Can you copy paste your almost working code here with the actual signature and stuff? I can try a stub.

**Comment by tejashah88** [comment]
Sure. I'm busy right now, but I'll share it later today.

**Comment by tejashah88** [comment]
Here's the almost-passing test code. Good luck!

```javascript
test('pass', function(t) {
  console.log(new Date().toString());
  var mockRes = invokeMiddleware({
    headers: {
      signature: 'Qc8OuaGEHWeL/39XTEDYFbOCufYWpwi45rqmM2R4WaSEYcSXq+hUko/88wv48+6SPUiEddWSEEINJFAFV5auYZsnBzqCK+SO8mGNOGHmLYpcFuSEHI3eA3nDIEARrXTivqqbH/LCPJHc0tqNYr3yPZRIR2mYFndJOxgDNSOooZX+tp2GafHHsjjShCjmePaLxJiGG1DmrL6fyOJoLrzc0olUxLmnJviS6Q5wBir899TMEZ/zX+aiBTt/khVvwIh+hI/PZsRq/pQw4WAvQz1bcnGNamvMA/TKSJtR0elJP+TgCqbVoYisDgQXkhi8/wonkLhs68pN+TurbR7GyC1vxw==',
      signaturecertchainurl: 'https://s3.amazonaws.com/echo.api/echo-api-cert-4.pem'
    },
    body: JSON.stringify({
      "version": "1.0",
      "session": {
        "new": true,
        "sessionId": "SessionId.7745e45d-3042-45eb-8e86-cab2cf285daf",
        "application": {
          "applicationId": "amzn1.ask.skill.75c997b8-610f-4eb4-bf2e-95810e15fba2"
        },
        "attributes": {},
        "user": {
          "userId": "amzn1.ask.account.AF6Z7574YHBQCNNTJK45QROUSCUJEHIYAHZRP35FVU673VDGDKV4PH2M52PX4XWGCSYDM66B6SKEEFJN6RYWN7EME3FKASDIG7DPNGFFFNTN4ZT6B64IIZKSNTXQXEMVBXMA7J3FN3ERT2A4EDYFUYMGM4NSQU4RTAQOZWDD2J7JH6P2ROP2A6QEGLNLZDXNZU2DL7BKGCVLMNA"
        }
      },
      "request": {
        "type": "IntentRequest",
        "requestId": "EdwRequestId.fa7428b7-75d0-44c8-aebb-4c222ed48ebe",
        "timestamp": "2017-02-10T07:27:59Z", // TROUBLE
        "locale": "en-US",
        "intent": {
          "name": "HelloWorld"
        },
        "inDialog": false
      }
    })
  })

  setTimeout(function() {
    t.equal(mockRes.statusCode, 200);
    console.log(mockRes._getData())
    /*t.deepEqual(JSON.parse(mockRes._getData()), {
      reason: 'certificate verification failed',
      status: 'failure'
    });*/
    t.end();
  }, 2000);
})
```
